File: internal/utils/Mapper/normalizeTimeValues.go
package mapper

import (
	"time"
)

// Lista över tidsformat vi vill stödja
var timeFormats = []string{
	time.RFC3339Nano,                // 2025-12-30T09:31:29.214605Z
	time.RFC3339,                    // 2025-12-30T09:31:29Z
	"2006-01-02 15:04:05",           // vanlig SQL datetime
	"2006-01-02 15:04:05.999999-07", // Postgres timestamptz
}

// Försök parse:a en sträng med alla formats i listan
func parseTimeGeneric(v string) (time.Time, bool) {
	for _, f := range timeFormats {
		t, err := time.Parse(f, v)
		if err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

// NormalizeTimeValues recursively traverses an unmarshalled JSON object
// and converts all valid time strings into time.Time
func NormalizeTimeValues(data interface{}) interface{} {
	switch v := data.(type) {
	case string:
		// Om det är en tidssträng → konvertera till time.Time
		if t, ok := parseTimeGeneric(v); ok {
			return t
		}
		return v
	case map[string]interface{}:
		// Gå igenom alla keys rekursivt
		for key, val := range v {
			v[key] = NormalizeTimeValues(val)
		}
		return v
	case []interface{}:
		// Gå igenom alla element i arrayen rekursivt
		for i, val := range v {
			v[i] = NormalizeTimeValues(val)
		}
		return v
	default:
		// Annars lämna värdet som det är
		return v
	}
}

File: src/utils/mapper/normalize-time-values.ts
// Try to parse a string as a date
function parseTimeGeneric(v: string): Date | null {
  // Using Date.parse which handles most common formats
  const parsed = new Date(v);
  if (!isNaN(parsed.getTime())) {
    return parsed;
  }
  return null;
}

// NormalizeTimeValues recursively traverses an unmarshalled JSON object
// and converts all valid time strings into Date objects
export function normalizeTimeValues(data: any): any {
  if (typeof data === "string") {
    // If it's a time string → convert to Date
    const parsed = parseTimeGeneric(data);
    if (parsed) {
      return parsed;
    }
    return data;
  }

  if (Array.isArray(data)) {
    // Go through all elements in the array recursively
    return data.map((val) => normalizeTimeValues(val));
  }

  if (data && typeof data === "object") {
    // Go through all keys recursively
    const result: Record<string, any> = {};
    for (const key in data) {
      if (Object.prototype.hasOwnProperty.call(data, key)) {
        result[key] = normalizeTimeValues(data[key]);
      }
    }
    return result;
  }

  // Otherwise leave the value as is
  return data;
}
